test/test_autocast.py

```diff
@@ -343,6 +343,22 @@ def test_cache_disabled(self):
         finally:
             torch._C._set_cached_tensors_enabled(False)
 
+    def test_autocast_prioritize(self):
```

Comment:
Could you add a small note explaining what this is testing? My current understanding is that:

index_put under amp has a cast policy of "promote", which seems to mean that (1) if any argument seen is float, then all the arguments are promoted to float, (2) double is ignored, (3) rest of the arguments must be precisely equal to the same dtype as the amp dtype. 
And because the amp dtype is not preserved across thread boundaries during autograd, it defaults to float16, and leads to the error because backward is being run in bfloat16 which is neither float/double and does not match the amp dtype of float16.